dali/test/python/test_operator_get_property.py

```diff
@@ -0,0 +1,138 @@
+# limitations under the License.
+
+from nvidia.dali import pipeline_def
+import os
+import numpy as np
+from nvidia.dali import fn
+from test_utils import get_dali_extra_path
+
+test_data_root = get_dali_extra_path()
+
+
+def _uint8_tensor_to_string(t):
+    return np.array(t).tobytes().decode()
+
+
+@pipeline_def
+def file_properties(root_path, device):
+    read, _ = fn.readers.file(file_root=root_path)
+    if device == 'gpu':
+        read = read.gpu()
+    return fn.get_property(read, key="source_info")
+
+
+def _test_file_properties(device):
+    root_path = os.path.join(test_data_root, 'db', 'single', 'png')
+    ref_paths = ["0/cat-1046544_640.png", "0/cat-111793_640.png", "0/cat-1245673_640.png", "0/cat-2184682_640.png",
+                 "0/cat-300572_640.png", "0/cat-3113513_640.png", "0/cat-3449999_640.png", "0/cat-3504008_640.png"]
+    p = file_properties(root_path, device, batch_size=8, num_threads=4, device_id=0)
+    p.build()
+    output = p.run()
+    for out in output:
+        out = out if device == 'cpu' else out.as_cpu()
+        for source_info, ref in zip(out, ref_paths):
+            assert _uint8_tensor_to_string(source_info) == ref
+
+
+def test_file_properties():
+    for dev in ['cpu', 'gpu']:
+        yield _test_file_properties, dev
+
+
+@pipeline_def
+def wds_properties(root_path, device):
+    read = fn.readers.webdataset(paths=[root_path], ext=['jpg'])
+    if device == 'gpu':
+        read = read.gpu()
+    return fn.get_property(read, key="source_info")
+
+
+def _test_wds_properties(device):
+    root_path = os.path.join(get_dali_extra_path(), "db/webdataset/MNIST/devel-0.tar")
+    ref_offset = [1536, 4096, 6144, 8704, 11264, 13824, 16384, 18432]
+    p = wds_properties(root_path, device, batch_size=8, num_threads=4, device_id=0)
+    p.build()
+    output = p.run()
+    for out in output:
+        out = out if device == 'cpu' else out.as_cpu()
+        for source_info, offset in zip(out, ref_offset):
+            assert _uint8_tensor_to_string(
+                source_info) == f"archive {root_path}tar file at \"{root_path}\"component offset {offset}"
+
+
+def test_wds_properties():
+    for dev in ['cpu', 'gpu']:
+        yield _test_wds_properties, dev
+
+
+@pipeline_def
+def tfr_properties(root_path, index_path, device):
+    import nvidia.dali.tfrecord as tfrec
+    inputs = fn.readers.tfrecord(path=root_path, index_path=index_path,
+                                 features={"image/encoded": tfrec.FixedLenFeature((), tfrec.string, ""),
+                                           "image/class/label": tfrec.FixedLenFeature([1], tfrec.int64, -1)})
+    if device == 'gpu':
+        return fn.get_property(inputs["image/encoded"], key="source_info").gpu(), \
+               fn.get_property(inputs["image/class/label"], key="source_info").gpu()
+    return fn.get_property(inputs["image/encoded"], key="source_info"), \
+           fn.get_property(inputs["image/class/label"], key="source_info")
```

Comment:
```suggestion
    out1 = fn.get_property(inputs["image/encoded"], key="source_info")
    out2 = fn.get_property(inputs["image/class/label"], key="source_info")
    if device == 'gpu':
      out1 = out1.gpu()
      out2 = out2.gpu()
    return out1, out2
```
^^ to avoid repetition

Reply:
The version with `.gpu()` at the end showed already a bug (or maybe not a bug, but a problem) in `Copy` function - see the ugly C-style cast. I'd keep the repetition to be guarded from that problem.

Reply:
I don't quite understand what was the problem, but the code that I am suggesting is equivalent